contracts/hydro/src/utils.rs

```diff
@@ -0,0 +1,290 @@
+    deps: &Deps<NeutronQuery>,
+    timestamp: Timestamp,
+) -> StdResult<(u64, Constants)> {
+    let current_constants: Vec<(u64, Constants)> = CONSTANTS
+        .range(
+            deps.storage,
+            None,
+            Some(Bound::inclusive(timestamp.nanos())),
+            Order::Descending,
+        )
+        .take(1)
+        .filter_map(|constants| match constants {
+            Ok(constants) => Some(constants),
+            Err(_) => None,
+        })
+        .collect();
+
+    Ok(match current_constants.len() {
+        1 => current_constants[0].clone(),
+        _ => {
+            return Err(StdError::generic_err(
+                "Failed to load constants active at the given timestamp.",
+            ));
+        }
+    })
+}
+
+// This function validates if user should be allowed to lock more tokens, depending on the total amount of
+// currently locked tokens, existence of the extra cap and the number of tokens user already locked in that cap.
+// Caps that we consider:
+//      1) Total cap: the maximum number of tokens that can be locked in the contract, regardless of the
+//         extra cap existence.
+//      2) Extra cap: the number of tokens reserved for users that had some voting power in previous round.
+//         It will be available to those users to lock additional tokens even if the public cap is filled.
+//         The extra cap will be available only for certain time frame at the beginning of the round. Duration
+//         will be determined through governance. After the extra cap duration has expired, any tokens left up
+//         to the total cap are allowed to be locked by any Hydro user.
+//      3) Public cap: during the existence of extra cap, new Hydro users will be allowed to lock tokens only
+//         in public cap, where public_cap = total_cap - extra_cap.
+//         After the extra cap duration expires, public cap becomes equal to the total cap.
+pub fn validate_locked_tokens_caps(
+    deps: &DepsMut<NeutronQuery>,
+    constants: &Constants,
+    current_round: u64,
+    sender: &Addr,
+    total_locked_tokens: u128,
+    amount_to_lock: u128,
+) -> Result<LockingInfo, ContractError> {
+    let lock_limit_reached_error = Err(ContractError::Std(StdError::generic_err(
+        "The limit for locking tokens has been reached. No more tokens can be locked.",
+    )));
+
+    // Regardless of public_cap and extra_cap, we must make sure that nobody can lock more than total_cap
+    let total_locked_after = total_locked_tokens + amount_to_lock;
+    if total_locked_after > constants.max_locked_tokens {
+        return lock_limit_reached_error;
+    }
+
+    let public_cap = constants.max_locked_tokens - constants.current_users_extra_cap;
+
+    // This branch will be executed in one of the following cases:
+    //   1) constants.current_users_extra_cap != 0 and there is SOME room in the public cap
+    //   2) constants.current_users_extra_cap == 0 and there is ENOUGH room in the public_cap.
```

Comment:
why do you list the special case here? isn't this branch just always executed if there is *some* room in the public cap, and then the check public_cap > total_locked_after handles the case where there is enough room in the public cap?

Reply:
Because if you look at the code, you might wonder what happens if `public_cap >= total_locked_after` gives `false` and you try to lock in extra_cap while it is not active (`constants.current_users_extra_cap == 0`), which shouldn't be allowed (even though there is also a check for this in can_user_lock_in_extra_cap()). It is just a clarification since the upper condition in that case becomes ` if total_cap >= total_locked_after`, which must always be true, otherwise we wouldn't be executing that line at all. Anyways, it's just something I thought could be useful to document, but I can remove it if you disagree.

Reply:
I found the comment a bit confusing, but let's leave it if you think it's helpful